File: python/No209.py
# -*- encoding: utf-8 -*-
'''
@Time    :   2022/04/20 23:06:47
@Author  :   Yu Runshen 
'''

# here put the import lib
import sys, os

class Solution:
    def minSubArrayLen(self, target, nums):
        result = sys.maxsize
        i = 0
        sum = 0
        
        for j in range(len(nums)):
            sum += nums[j]
            while sum >= target:
                length = j - i + 1
                result = min(result, length)
                sum -= nums[i]
                i += 1
        return 0 if result==sys.maxsize else result

s = Solution()
print(s.minSubArrayLen(target = 7, nums = [2,3,1,2,4,3]))

File: c++/No209.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int size = nums.size();
        int i = 0;
        int j = 0;
        int sum = 0;
        int length = size + 1;

        while (j < size)
        {
            sum += nums[j];
            while (sum >= target) {
                int tmp_len = j - i + 1;
                length = length < tmp_len ? length : tmp_len;
                sum -= nums[i];
                i += 1;
            }
            j += 1;
        }
        return length==(size+1) ? 0 : length;
    }
};